### PR TITLE
feat(#118): improve error message for `--execute`

### LINKED ISSUE #118: improve error message for `-x`
if there's an error when we execute a command, ideally we should show that in our error message
image

was actually a "file not found" error for the go program I was trying to run

### PR DESCRIPTION
**Pull request by AlejandroSuero**
With these changes, the error messages will show error that occurred. Closes #118.

Screenshot 2024-09-04 at 21 23 47

> [!note]
> As you can see, it only shows the first line of the message.

---

**Possible improvements**:
- Show the full message, current state:

Screenshot 2024-09-04 at 21 30 51

```txt
$ git defer
git: 'defer' is not a git command. See 'git --help'.

The most similar commands are
        rerere
        revert
```

### CODE REVIEW

**Comment by bashbunni** [comment]
Hey @AlejandroSuero thank you so much for this contribution! Do you know why the full error message isn't displayed at the moment? This example output looks a bit rough
image

That said, I didn't reproduce the garbled output, so looks good??
image

Let me know if I'm missing anything :)

**Comment by AlejandroSuero** [comment]
@bashbunni The code I committed has the version where it prints the first line of the error, because I ran into the problem with the garbled output.

I tested printing out to the terminal with `fmt.Println` not using `errorDetails` styles and it prints fine:

Screenshot 2024-09-05 at 21 02 12

Maybe the error is in the way **lipgloss** is rendering it, I will look into it and comeback.

**Comment by AlejandroSuero** [comment]
@bashbunni as I thought, the problem is the way lipgloss is rendering it.

I'm kinda new to **lipgloss** so I don't know the specifics, but it seems that is calling `Render` on the whole string which for some reason causes problems.

I arrived to the solution of splitting the string and rendering it line by line like:

changes on the error rendering

Which results in the following output:

Screenshot 2024-09-05 at 21 32 05

Do you want me to commit it or is there a better way to achieve this with **lipgloss**?

File: main.go [hunk 1]
```diff
@@ -85,6 +85,10 @@ func main() {
 	if config.Execute != "" {
 		input, err = executeCommand(config)
 		if err != nil {
+			if input != "" {
+				// show the full error message output
+				printErrorFatal("Something went wrong", errors.New(input))
```
Comment (ccoVeille) [inline main.go:90]: What about replacing the line by this

```suggestion
				err = fmt.Errorf("%w %s", err, input)
```

File: main.go [hunk 2]
```diff
@@ -85,6 +85,10 @@ func main() {
 	if config.Execute != "" {
 		input, err = executeCommand(config)
```
Comment (ccoVeille) [inline main.go:86]: Is it an input or an output?

Comment (AlejandroSuero) [reply main.go:86]: In the main branch is how it was:

https://github.com/charmbracelet/freeze/blob/335467362a96eec3e0ff6514a27006dfccaa7b62/main.go#L85-L93

Do I change it to be `output` variable?

Comment (ccoVeille) [reply main.go:86]: I think so, look at execute method, the variable is named output, no?

Comment (AlejandroSuero) [reply main.go:86]: Yeah, it's called `out`, so we can call it that. After all, it exits once `printErrorFatal` call ends.

File: error.go [hunk 3]
```diff
@@ -3,18 +3,20 @@ package main
...
-)
+var errorHeader = lipgloss.NewStyle().Foreground(lipgloss.Color("#F1F1F1")).Background(lipgloss.Color("#FF5F87")).Bold(true).Padding(0, 1).Margin(1).MarginLeft(2).SetString("ERROR")
+var errorDetails = lipgloss.NewStyle().Foreground(lipgloss.Color("#757575")).MarginLeft(2)
 
 func printError(title string, err error) {
 	fmt.Printf("%s\n", lipgloss.JoinHorizontal(lipgloss.Center, errorHeader.String(), title))
-	fmt.Printf("%s\n", errorDetails.Render(err.Error()))
+	splittedError := strings.Split(err.Error(), "\n")
+	for _, line := range splittedError {
+		fmt.Printf("%s\n", errorDetails.Render(line))
```
[the remark this answers is not in the document]
Comment (AlejandroSuero) [reply error.go:18]: Yeah I did `Printf` to match the one above and how it was before. Should I change the one above too (displays the error title)?

**Review by ccoVeille — APPROVED** [review]

**Comment by bashbunni** [comment]
Sorry for the delay here, but I want to spend more time with that underlying issue in Lip Gloss before merging this. Ideally we can include the entire error message here with Lip Gloss styling :pray:

**Comment by bashbunni** [comment]
Hey update on this. I'm working to debug this in Lip Gloss and while I can reproduce the issue, for some reason the output is wrong, but the string even provided to `fmt.Print()` is correctly formatted and has all the content (or so it seems). I copied that string value to a file (included in the linked commit) then used `less -R printf-val.txt` which gave the expected output (not sure if the VS Code debugger includes newlines though. There are no newlines in there from what I saw.

https://github.com/charmbracelet/freeze/commit/4d15621fbe6ad35996d74699e0e33908448f58d6

This is the result of `less -R printf-val.txt`
image

This is what prints to `stdout`
image

I'm working on this branch to debug this issue: https://github.com/charmbracelet/freeze/tree/bunni-improve-error-msg

I have a branch where the output formats correctly, but the solution is too hacky for my liking. I'd rather figure out why this is happening in Lip Gloss. https://github.com/charmbracelet/freeze/commit/fdcd0de234f72494adc420224bc5ae11f43142b8

To debug in my debugger I'm running `dlv debug --headless --api-version=2 --listen=127.0.0.1:43000 . -- -x "git defer" -c full` and using the VS Code debugger with the following `launch.json`

```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387

    // dlv debug --headless --api-version=2 --listen=127.0.0.1:43000 . -- -x "git defer" -c full
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Connect to server",
            "type": "go",
            "request": "attach",
            "mode": "remote",
            "remotePath": "${workspaceFolder}",
            "port": 43000,
            "host": "127.0.0.1"
        }

    ]
}
```

**Comment by bashbunni** [comment]
the issue is fixed when I comment out line 65 (`l += s` in the `default` case) of lipgloss `master` `align.go`. Once that is fixed on lipgloss then we can merge this :)

**Comment by bashbunni** [comment]
Found the issue in Lip Gloss, have an open PR. Will take another look at this once that's merged https://github.com/charmbracelet/lipgloss/pull/386

**Comment by AlejandroSuero** [comment]
Thanks @bashbunni as I said on the Discord, I was a bit busy with take homes and got the job and started the next day after telling me that I got it 😄 , and now with all the onboarding and catching up, couldn't debug it myself. Good job finding the error 👍

**Comment by bashbunni** [comment]
@AlejandroSuero no problem at all, you've already helped us lots with this PR + highlighting this issue in the first place. I hope the new gig is treating you well!

File: main.go (hunk 2 above)
Comment (bashbunni) [reply main.go:86]: Just to update on why I changed this back to `input` - it's typically an input file that's being provided to the program whose value is checked further down, so it must be set. If this is renamed, the tests fail and we get an error for all `--execute` commands that don't get an error back. For example:

image

**Comment by bashbunni** [comment]
@AlejandroSuero Thank you for your work on this! 🚀

**Pull request merged**